File: Data Structures & Algorithms/three-integer-sum/submission-5.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
       sort(nums.begin(), nums.end());
        int arrSize = nums.size();
        vector<vector<int>> ans;
        for(int i = 0; i < arrSize-2;i++){
            while(i > 0 and nums[i] == nums[i-1] and i < arrSize - 2){
                i++;
            }

            int j = i+1;
            int k = arrSize-1;
            while (j < k){
                int sum = nums[i] + nums[j] + nums[k];
                if( sum == 0){
                    ans.push_back({nums[i], nums[j], nums[k]});
                    j ++;
                    k --;
                    while(nums[j] == nums[j-1] and j < k){
                        j++;
                    }
                }
                else if(sum < 0){
                    j ++;
                }
                else{
                    k --;
                }
                
                // while(k < arrSize - 1 and nums[k] == nums[k+1] and k > j){
                //     k --;
                // }
            }

        }
        return ans;
    }
};

File: Data Structures & Algorithms/top-k-elements-in-list/submission-5.ts
class Solution {
    /**
     * @param {number[]} nums
     * @param {number} k
     * @return {number[]}
     */
    topKFrequent(nums: number[], k: number): number[] {
        // 1. Explicitly type the accumulator as a Record<number, number>
        const count = nums.reduce<Record<number, number>>((acc, cur) => {
            acc[cur] = acc[cur] !== undefined ? acc[cur] + 1 : 1;
            return acc;
        }, {});

        // 2. Object.entries automatically provides [string, number], no need to remap it
        const buc_sort = Object.entries(count).sort((a, b) => b[1] - a[1]);
        const ans: number[] = [];

        // 3. Prevent runtime errors if k is greater than unique elements
        const limit = Math.min(k, buc_sort.length);
        for (let i = 0; i < limit; i++) {
            ans.push(Number(buc_sort[i][0]));
        }

        return ans;
    }
}
